File: src/classes/Reports.java
package src.classes;
import src.Election;


public class Reports {
	// attributes
	private Election electionInfo = null;


	// constructor method
	public Reports(Election election) {
		this.electionInfo = election;
	}


	// methods
	public void displayNumberOfElectedCandidates() {
		System.out.println("Número de vagas: " + electionInfo.getNumberOfElectedCandidates());
		System.out.println();
	}

	public void displayElectedCandidates() {
		System.out.println("Vereadores eleitos:");
		for (int i = 0; i < electionInfo.getElectedCandidates().size(); i++) {
			Candidate candidate = electionInfo.getElectedCandidates().get(i);
			String votesOnSingularOrPlural = " votos)";
			if (candidate.getVotes() <= 0) {
				votesOnSingularOrPlural = " voto)";
			}
			System.out.println(
				i+1 + " - " +
				candidate.getName() + " / " + candidate.getBallotBoxAlias() +
				" (" +
				candidate.getPoliticalPartyAliasByNumber(
					electionInfo.getParties()
				)
				+ ", " + candidate.getVotes() + votesOnSingularOrPlural
			);
		}
		System.out.println();
	}

	public void displayMostVotedCandidates() {
		System.out.println("Candidatos mais votados (em ordem decrescente de votação e respeitando número de vagas):");
		for (Candidate candidate : electionInfo.getMostVotedCandidates()) {
			String votesOnSingularOrPlural = " votos)";
			if (candidate.getVotes() <= 0) {
				votesOnSingularOrPlural = " voto)";
			}
			System.out.println(
				candidate.getPosition() + " - " +
				candidate.getName() + " / " + candidate.getBallotBoxAlias() +
				" (" +
				candidate.getPoliticalPartyAliasByNumber(
					electionInfo.getParties()
				)
				+ ", " + candidate.getVotes() + votesOnSingularOrPlural
			);
		}
		System.out.println();
	}

	public void displayMostVotedAndNotElectedCandidates() {
		System.out.println("Teriam sido eleitos se a votação fosse majoritária, e não foram eleitos:\n(com sua posição no ranking de mais votados)");
		for (Candidate candidate : electionInfo.getMostVotedAndNotElectedCandidates()) {
			String votesOnSingularOrPlural = " votos)";
			if (candidate.getVotes() <= 0) {
				votesOnSingularOrPlural = " voto)";
			}
			System.out.println(
				candidate.getPosition() + " - " +
				candidate.getName() + " / " + candidate.getBallotBoxAlias() +
				" (" +
				candidate.getPoliticalPartyAliasByNumber(
					electionInfo.getParties()
				)
				+ ", " + candidate.getVotes() + votesOnSingularOrPlural
			);
		}
		System.out.println();
	}

	public void displayElectedButNotMostVotedCandidates() {
		System.out.println("Eleitos, que se beneficiaram do sistema proporcional:\n(com sua posição no ranking de mais votados)");
		for (Candidate candidate : electionInfo.getElectedButNotMostVotedCandidates()) {
			String votesOnSingularOrPlural = " votos)";
			if (candidate.getVotes() <= 0) {
				votesOnSingularOrPlural = " voto)";
			}
			System.out.println(
				candidate.getPosition() + " - " +
				candidate.getName() + " / " + candidate.getBallotBoxAlias() +
				" (" +
				candidate.getPoliticalPartyAliasByNumber(
					electionInfo.getParties()
				)
				+ ", " + candidate.getVotes() + votesOnSingularOrPlural
			);
		}
		System.out.println();
	}

	public void displayPartiesTotalVotes() {
		System.out.println("Votação dos partidos e número de candidatos eleitos:");
		for (int i = 0; i < electionInfo.getPartiesOrderedByTotalVotes().size(); i++) {
			Party party = electionInfo.getPartiesOrderedByTotalVotes().get(i);
			String candidatesOnSingularOrPlural = " candidatos eleitos";
			String[] votesOnSingularOrPlural = {" votos", " nominais"};
			if (electionInfo.getElectedCandidatesNumberFromParty(party) <= 1) {
				candidatesOnSingularOrPlural = " candidato eleito";
			}
			if (party.getTotalVotes() <= 1) {
				votesOnSingularOrPlural = new String[] {" voto", " nominal"};
			}
			System.out.println(
				i+1 + " - " + party.getAlias() + " - " + party.getNumber() +
				", " + party.getTotalVotes() + votesOnSingularOrPlural[0] + " (" + (party.getTotalVotes() - party.getPartyVotes()) + votesOnSingularOrPlural[1] + " e " + party.getPartyVotes() + " de legenda), " +
				electionInfo.getElectedCandidatesNumberFromParty(party) + candidatesOnSingularOrPlural
			);
		}
		System.out.println();
	}

	public void displayPartiesVotes() {
		System.out.println("Votação dos partidos (apenas votos de legenda):");
		for (int i = 0; i < electionInfo.getPartiesOrderedByPartyVotes().size(); i++) {
			Party party = electionInfo.getPartiesOrderedByPartyVotes().get(i);
			if (party.getPartyVotes() <= 0) {
				System.out.printf("%d - %s - %d, %d voto de legenda (proporção não calculada, 0 voto no partido)\n",
					i+1, party.getAlias(), party.getNumber(), party.getPartyVotes()
				);
			}
			else {
				System.out.printf("%d - %s - %d, %d votos de legenda (%.2f%% do total do partido)\n",
					i+1, party.getAlias(), party.getNumber(), party.getPartyVotes(), electionInfo.getPartyVotesPercent(party)
				);
			}
		}
		System.out.println();
	}

	public void displayPartiesCandidates() {
		System.out.println("Primeiro e último colocados de cada partido:");
		for (int i = 0; i < electionInfo.getPartiesWithCandidatesOrderedByVotes().size(); i++) {
			Party party = electionInfo.getPartiesWithCandidatesOrderedByVotes().get(i);
			Candidate firstCandidate = party.getCandidates().get(0);
			Candidate lastCandidate = party.getCandidates().get(party.getCandidates().size()-1);
			for (Candidate actualCandidate : party.getCandidates()) {
				if (actualCandidate.getVotesDestiny().equals("Válido")) {
					firstCandidate = actualCandidate;
					break;
				}
			}
			for (int k = party.getCandidates().size() -1; k >= 0; k--) {
				Candidate actualCandidate = party.getCandidates().get(k);
				if (actualCandidate.getVotesDestiny().equals("Válido")) {
					lastCandidate = actualCandidate;
					break;
				}
			}
			if (!firstCandidate.getVotesDestiny().equals("Válido") && !lastCandidate.getVotesDestiny().equals("Válido")) {
				continue;
			}
			String firstCandidateVotesOnSingularOrPlural = " votos) / ";
			if (firstCandidate.getVotes() <= 1) {
				firstCandidateVotesOnSingularOrPlural = " voto) / ";
			}
			String lastCandidateVotesOnSingularOrPlural = " votos)";
			if (lastCandidate.getVotes() <= 1) {
				lastCandidateVotesOnSingularOrPlural = " voto)";
			}
			System.out.println(i+1 + " - " + party.getAlias() + " - " + party.getNumber() + ", " +
				firstCandidate.getBallotBoxAlias() + " (" + firstCandidate.getNumber() + ", " + firstCandidate.getVotes() + firstCandidateVotesOnSingularOrPlural +
				lastCandidate.getBallotBoxAlias() + " (" + lastCandidate.getNumber() + ", " + lastCandidate.getVotes() + lastCandidateVotesOnSingularOrPlural
			);
		}
		System.out.println();
	}

	public void displayAgeOfElectedCandidates() {
		System.out.println("Eleitos, por faixa etária (na data da eleição):");
		Utils utils = new Utils();
		int age1 = 0, age2 = 0, age3 = 0, age4 = 0, age5 = 0, ageTotal = 0;
		for (Candidate candidate : electionInfo.getElectedCandidates()) {
			int age = utils.calculateAge(
				electionInfo.parseDateToString(candidate.getBirthDate()),
				electionInfo.parseDateToString(electionInfo.getElectionDate())
			);
			if (age < 30) {
				age1++;
			}
			else if (30 <= age && age < 40) {
				age2++;
			}
			else if (40 <= age && age < 50) {
				age3++;
			}
			else if (50 <= age && age < 60) {
				age4++;
			}
			else {
				age5++;
			}
			ageTotal++;
		}
		System.out.print("      Idade < 30: " + age1);
		System.out.printf(" (%.2f%%)\n", ((float) age1 / ageTotal * 100));
		System.out.print("30 <= Idade < 40: " + age2);
		System.out.printf(" (%.2f%%)\n", ((float) age2 / ageTotal * 100));
		System.out.print("40 <= Idade < 50: " + age3);
		System.out.printf(" (%.2f%%)\n", ((float) age3 / ageTotal * 100));
		System.out.print("50 <= Idade < 60: " + age4);
		System.out.printf(" (%.2f%%)\n", ((float) age4 / ageTotal * 100));
		System.out.print("60 <= Idade : " + age5);
		System.out.printf(" (%.2f%%)\n", ((float) age5 / ageTotal * 100));
		System.out.println();
	}

	public void displayGenderOfElectedCandidates() {
		System.out.println("Eleitos, por sexo:");
		int male = 0, female = 0, total = 0;
		for (Candidate candidate : electionInfo.getElectedCandidates()) {
			if (candidate.getGender() == 'M') {
				male++;
			}
			else {
				female++;
			}
			total++;
		}
		System.out.print("Feminino: "+ female);
		System.out.printf(" (%.2f%%)\n", electionInfo.getGeneralPercent(female, total));
		System.out.print("Masculino: " + male);
		System.out.printf(" (%.2f%%)\n", electionInfo.getGeneralPercent(male, total));
		System.out.println();
	}

	public void displayAllVotes() {
		System.out.printf(
			"Total de votos válidos:    %d\n",
			electionInfo.getAmountOfTotalVotes()
		);
		System.out.printf(
			"Total de votos nominais:   %d (%.2f%%)\n",
			electionInfo.getAmountOfCandidatesVotes(),
			electionInfo.getGeneralPercent(
				electionInfo.getAmountOfCandidatesVotes(), electionInfo.getAmountOfTotalVotes()
			)
		);
		System.out.printf(
			"Total de votos de legenda: %d (%.2f%%)",
			electionInfo.getAmountOfPartyVotes(),
			electionInfo.getGeneralPercent(
				electionInfo.getAmountOfPartyVotes(), electionInfo.getAmountOfTotalVotes()
			)
		);
	}

}

File: source/reports.cpp
#include "../include/reports.h"


Reports::Reports(Election &election) {
	this->electionInfo = election;
}
Reports::~Reports() {
	//// do nothing
}


const string& Reports::getPoliticalPartyNameByNumber(vector<Party*> parties, int politicalPartyNumber) {
	const string& defaultStr = "";

	for (Party* party : parties) {
		if (party->getNumber() == politicalPartyNumber) {
			return party->getName();
		}
	}
	return defaultStr;
}
const string& Reports::getPoliticalPartyAliasByNumber(vector<Party*> parties, int politicalPartyNumber) {
	const string& defaultStr = "";

	for (Party* party : parties) {
		if (party->getNumber() == politicalPartyNumber) {
			return party->getAlias();
		}
	}
	return defaultStr;
}

void Reports::displayNumberOfElectedCandidates() {
	cout << "Número de vagas: " << electionInfo.getNumberOfElectedCandidates() << endl;
	cout << endl;
}

void Reports::displayElectedCandidates() {
	cout << "Vereadores eleitos:" << endl;
	for (int i = 0; i < (int)electionInfo.getElectedCandidates().size(); i++) {
		Candidate* candidate = electionInfo.getElectedCandidates()[i];
		string votesOnSingularOrPlural = " votos)";
		if (candidate->getVotes() <= 0) {
			votesOnSingularOrPlural = " voto)";
		}
		cout << i+1 << " - " << candidate->getName() << " / " << candidate->getBallotBoxAlias() <<	" (" << getPoliticalPartyAliasByNumber(electionInfo.getParties(), candidate->getPoliticalPartyNumber()) << ", " << candidate->getVotes() << votesOnSingularOrPlural << endl;
	}
	cout << endl;
}

void Reports::displayMostVotedCandidates() {
	cout << "Candidatos mais votados (em ordem decrescente de votação e respeitando número de vagas):" << endl;
	for (Candidate* candidate : electionInfo.getMostVotedCandidates()) {
		string votesOnSingularOrPlural = " votos)";
		if (candidate->getVotes() <= 0) {
			votesOnSingularOrPlural = " voto)";
		}
		cout << candidate->getPosition() << " - " << candidate->getName() << " / " << candidate->getBallotBoxAlias() << " (" << getPoliticalPartyAliasByNumber(electionInfo.getParties(), candidate->getPoliticalPartyNumber()) << ", " << candidate->getVotes() << votesOnSingularOrPlural << endl;
	}
	cout << endl;
}

void Reports::displayMostVotedAndNotElectedCandidates() {
	cout << "Teriam sido eleitos se a votação fosse majoritária, e não foram eleitos:\n(com sua posição no ranking de mais votados)" << endl;
	for (Candidate* candidate : electionInfo.getMostVotedAndNotElectedCandidates()) {
		string votesOnSingularOrPlural = " votos)";
		if (candidate->getVotes() <= 0) {
			votesOnSingularOrPlural = " voto)";
		}
		cout << candidate->getPosition() << " - " << candidate->getName() << " / " << candidate->getBallotBoxAlias() << " (" << getPoliticalPartyAliasByNumber(electionInfo.getParties(), candidate->getPoliticalPartyNumber()) << ", " << candidate->getVotes() << votesOnSingularOrPlural << endl;
	}
	cout << endl;
}

void Reports::displayElectedButNotMostVotedCandidates() {
	cout << "Eleitos, que se beneficiaram do sistema proporcional:\n(com sua posição no ranking de mais votados)" << endl;
	for (Candidate* candidate : electionInfo.getElectedButNotMostVotedCandidates()) {
		string votesOnSingularOrPlural = " votos)";
		if (candidate->getVotes() <= 0) {
			votesOnSingularOrPlural = " voto)";
		}
		cout << candidate->getPosition() << " - " << candidate->getName() << " / " << candidate->getBallotBoxAlias() << " (" << getPoliticalPartyAliasByNumber(electionInfo.getParties(), candidate->getPoliticalPartyNumber()) << ", " << candidate->getVotes() << votesOnSingularOrPlural << endl;
	}
	cout << endl;
}

void Reports::displayPartiesTotalVotes() {
	cout << "Votação dos partidos e número de candidatos eleitos:" << endl;
	for (int i = 0; i < (int)electionInfo.getPartiesOrderedByTotalVotes().size(); i++) {
		Party* party = electionInfo.getPartiesOrderedByTotalVotes()[i];
		string candidatesOnSingularOrPlural = " candidatos eleitos";
		vector<string> votesOnSingularOrPlural = {" votos", " nominais"};
		if (electionInfo.getElectedCandidatesNumberFromParty(*party) <= 1) {
			candidatesOnSingularOrPlural = " candidato eleito";
		}
		if (party->getTotalVotes() <= 1) {
			vector<string> votesOnSingularOrPlural = {" voto", " nominal"};
		}
		cout << i+1 << " - " << party->getAlias() << " - " << party->getNumber() << ", " << party->getTotalVotes() << votesOnSingularOrPlural[0] << " (" << (party->getTotalVotes() - party->getPartyVotes()) << votesOnSingularOrPlural[1] << " e " << party->getPartyVotes() << " de legenda), " << electionInfo.getElectedCandidatesNumberFromParty(*party) << candidatesOnSingularOrPlural << endl;
	}
	cout << endl;
}

void Reports::displayPartiesVotes() {
	cout << fixed;
	cout << setprecision(2);
	cout << "Votação dos partidos (apenas votos de legenda):" << endl;
	for (int i = 0; i < (int)electionInfo.getPartiesOrderedByPartyVotes().size(); i++) {
		Party* party = electionInfo.getPartiesOrderedByPartyVotes()[i];
		if (party->getPartyVotes() <= 0) {
			cout << i+1 << " - " << party->getAlias() << " - " << party->getNumber() << ", " << party->getPartyVotes() << " voto de legenda (proporção não calculada, 0 voto no partido)" << endl;
		}
		else {
			cout << i+1 << " - " << party->getAlias() << " - " << party->getNumber() << ", " << party->getPartyVotes() << " votos de legenda (" << electionInfo.getPartyVotesPercent(*party) << "%% do total do partido)" << endl;
		}
	}
	cout << endl;
}

void Reports::displayPartiesCandidates() {
	cout << "Primeiro e último colocados de cada partido:" << endl;
	for (int i = 0; i < (int)electionInfo.getPartiesWithCandidatesOrderedByVotes().size(); i++) {
		Party* party = electionInfo.getPartiesWithCandidatesOrderedByVotes()[i];
		Candidate* firstCandidate = party->getCandidates()[0];
		Candidate* lastCandidate = party->getCandidates()[(party->getCandidates().size()-1)];
		for (Candidate* actualCandidate : party->getCandidates()) {
			if (actualCandidate->getVotesDestiny() == "Válido") {
				firstCandidate = actualCandidate;
				break;
			}
		}
		for (int k = party->getCandidates().size() -1; k >= 0; k--) {
			Candidate* actualCandidate = party->getCandidates()[k];
			if (actualCandidate->getVotesDestiny() == "Válido") {
				lastCandidate = actualCandidate;
				break;
			}
		}
		if (!(firstCandidate->getVotesDestiny() == "Válido") && !(lastCandidate->getVotesDestiny() == "Válido")) {
			continue;
		}
		string firstCandidateVotesOnSingularOrPlural = " votos) / ";
		if (firstCandidate->getVotes() <= 1) {
			firstCandidateVotesOnSingularOrPlural = " voto) / ";
		}
		string lastCandidateVotesOnSingularOrPlural = " votos)";
		if (lastCandidate->getVotes() <= 1) {
			lastCandidateVotesOnSingularOrPlural = " voto)";
		}
		cout << i+1 << " - " << party->getAlias() << " - " << party->getNumber() << ", " << firstCandidate->getBallotBoxAlias() << " (" << firstCandidate->getNumber() << ", " << firstCandidate->getVotes() << firstCandidateVotesOnSingularOrPlural << lastCandidate->getBallotBoxAlias() << " (" << lastCandidate->getNumber() << ", " << lastCandidate->getVotes() << lastCandidateVotesOnSingularOrPlural << endl;
	}
	cout << endl;
}

void Reports::displayAgeOfElectedCandidates() {
	cout << "Eleitos, por faixa etária (na data da eleição):" << endl;
	Utils utils = Utils();
	int age1 = 0, age2 = 0, age3 = 0, age4 = 0, age5 = 0, ageTotal = 0;
	for (Candidate* candidate : electionInfo.getElectedCandidates()) {
		int age = utils.calculateAge(
			candidate->getBirthDate(),
			electionInfo.getElectionDate()
		);
		if (age < 30) {
			age1++;
		}
		else if (30 <= age && age < 40) {
			age2++;
		}
		else if (40 <= age && age < 50) {
			age3++;
		}
		else if (50 <= age && age < 60) {
			age4++;
		}
		else {
			age5++;
		}
		ageTotal++;
	}
	cout << "      Idade < 30: " << age1 << " (" << ((float) age1 / ageTotal * 100) << "%%)" << endl;
	cout << "30 <= Idade < 40: " << age2 << " (" << ((float) age2 / ageTotal * 100) << "%%)" << endl;
	cout << "40 <= Idade < 50: " << age3 << " (" << ((float) age3 / ageTotal * 100) << "%%)" << endl;
	cout << "50 <= Idade < 60: " << age4 << " (" << ((float) age4 / ageTotal * 100) << "%%)" << endl;
	cout << "60 <= Idade     : " << age5 << " (" << ((float) age5 / ageTotal * 100) << "%%)" << endl;
	cout << endl;
}

void Reports::displayGenderOfElectedCandidates() {
	cout << "Eleitos, por sexo:" << endl;
	int male = 0, female = 0, total = 0;
	for (Candidate* candidate : electionInfo.getElectedCandidates()) {
		if (candidate->getGender() == 'M') {
			male++;
		}
		else {
			female++;
		}
		total++;
	}
	cout << "Feminino: " << female << " (" << electionInfo.getGeneralPercent(female, total) << "%%)" << endl;
	cout << "Masculino: " << male << " (" << electionInfo.getGeneralPercent(male, total) << "%%)" << endl;
	cout << endl;
}

void Reports::displayAllVotes() {
	cout << "Total de votos válidos:    " << electionInfo.getAmountOfTotalVotes() << endl;
	cout << "Total de votos nominais:   " << electionInfo.getAmountOfCandidatesVotes() << " (" << electionInfo.getGeneralPercent(electionInfo.getAmountOfCandidatesVotes(), electionInfo.getAmountOfTotalVotes()) << "%%)" << endl;
	cout << "Total de votos de legenda: " << electionInfo.getAmountOfPartyVotes() << " (" << electionInfo.getGeneralPercent(electionInfo.getAmountOfPartyVotes(), electionInfo.getAmountOfTotalVotes()) << "%%)" << endl;
}
